packages/bpk-component-infinite-scroll/src/withInfiniteScroll.js

```diff
@@ -0,0 +1,232 @@
+        }
+      };
+      this.handleSeeMoreClick = this.handleSeeMoreClick.bind(this);
+      this.handleKeyPress = this.handleKeyPress.bind(this);
+      this.observer = new IntersectionObserver(this.handleIntersection, {
+        threshold: 1,
+      });
+    }
+
+    componentDidMount() {
+      if (this.sentinel) {
+        this.observer.observe(this.sentinel);
+      }
+    }
+
+    componentDidUpdate() {
+      if (this.sentinel && this.state.index > 0) {
+        this.observer.observe(this.sentinel);
+      }
+    }
+
+    async loadMore() {
+      const { index, elementsToRender } = this.state;
+      const { elementsPerScroll, seeMoreAfter } = this.props;
+      const nextElements = await this.onItemsFetch(index, elementsPerScroll);
+      if (nextElements && nextElements.length > 0) {
+        const nextIndex = index + elementsPerScroll;
+        return {
+          index: nextIndex,
+          elementsToRender: elementsToRender.concat(nextElements),
+          showSeeMore: seeMoreAfter === index / elementsPerScroll,
+        };
+      }
+      this.props.onScrollFinished({
+        totalNumberElements: elementsToRender.length,
+      });
+      return {
+        isListFinished: true,
+      };
+    }
+
+    async handleSeeMoreClick(e: SyntheticEvent<HTMLButtonElement>) {
+      this.props.onClickSeeMore(e);
+      this.setState(await this.loadMore());
+    }
+
+    handleKeyPress(e: SyntheticKeyboardEvent<HTMLButtonElement>) {
+      const { key } = e;
+      if (key === 'Enter' || key === ' ') {
+        this.handleSeeMoreClick(e);
+      }
+    }
+
+    render() {
+      const { elementsToRender, isListFinished, showSeeMore } = this.state;
+      const { seeMoreComponent, loadingComponent } = this.props;
+
+      const rest = omit(this.props, [
+        'elementsPerScroll',
+        'loadingComponent',
+        'onItemsFetch',
+        'onClickSeeMore',
+        'onScroll',
+        'onScrollFinished',
+        'seeMoreComponent',
+        'seeMoreAfter',
+      ]);
+
+      let loadingOrButton = null;
+
+      if (!isListFinished) {
+        if (showSeeMore) {
+          loadingOrButton = (
+            <div
+              role="button"
+              tabIndex={0}
+              onKeyPress={this.handleKeyPress}
+              onClick={this.handleSeeMoreClick}
+            >
+              {seeMoreComponent}
+            </div>
+          );
+        } else {
+          loadingOrButton = (
+            <div
+              ref={spinner => {
+                this.sentinel = spinner;
+              }}
+              className={
+                loadingComponent ? null : getClassNames('bpk-sentinel')
+              }
+            >
+              {loadingComponent}
+            </div>
+          );
+        }
+      }
+
+      return (
+        <div>
+          <ComponentToExtend {...rest} elements={elementsToRender} />
+          {loadingOrButton}
+        </div>
+      );
+    }
+  }
+
+  Infinitize.propTypes = {
```

Comment:
can you move proptypes and default props to inside the class? like this: https://github.com/Skyscanner/backpack/blob/master/native/packages/react-native-bpk-component-progress/src/BpkProgress.js#L99

it works better with flow.